examples/websocket/src/test/kotlin/cnt/websocket/LectureChatWebSocketServerTest.kt

```diff
@@ -0,0 +1,146 @@
+                .containsExactly(
+                    "User Client1 joined",
+                    "You and 0 users are currently in this session",
+                    "Welcome to the lecture!",
+                    "Client1: Hello I like your book!"
+                )
+        }
+    }
+
+    @Test
+    fun `two user with same username try to start a session`() {
+        val session1 = connectToServer(client1)
+        val session2 = connectToServer(client1)
+        val recordedMessage = LectureChatTestClient.stopRecordingMessages(5)
+
+        session1.close()
+        session2.close()
+
+        assertThat(recordedMessage)
+            .hasSize(6)
+            .filteredOn { it != null }
+            .hasSize(5) // one null value removed to verify no more messages besides following
+            .extracting("message")
+            .map<String> { (it as String).substringAfterLast("-").removePrefix(" ") }
+            .containsExactlyInAnyOrder(
+                "User Client1 joined",
+                "User Client1 has already joined. Take another username",
+                "Session was closed",
+                "You and 0 users are currently in this session",
+                "Welcome to the lecture!",
+            )
+    }
+
+    @Test
+    fun `two different user starting a session - client1 sends a message - and after that one leaves`() {
+        val sessionClient1 = connectToServer(client1)
+        val sessionClient2 = connectToServer(client2)
+        sessionClient2.close()
+        val recordedMessage = LectureChatTestClient.stopRecordingMessages(5)
+
+        sessionClient1.close()
+        assertThat(recordedMessage)
+            .hasSize(6)
+            .filteredOn { it != null }
+            .hasSize(5) // one null value removed to verify no more messages besides following
+            .extracting("message")
+            .map<String> { (it as String).substringAfterLast("-").removePrefix(" ") }
+            .containsExactlyInAnyOrder(
+                "User Client1 joined",
+                "You and 0 users are currently in this session",
+                "Welcome to the lecture!",
+                "User Client2 joined",
+                "User Client2 left the session",
+            )
+    }
+
+    private fun connectToServer(uri: URI): Session {
+        return ContainerProvider.getWebSocketContainer().connectToServer(LectureChatTestClient::class.java, uri)
+    }
+
+    @ClientEndpoint(
+        encoders = [MessageEnDecoder::class],
+        decoders = [MessageEnDecoder::class]
+    )
+    class LectureChatTestClient {
+
+        @OnMessage
+        fun onMessage(msg: Message) {
+            println(msg.message)
+            message.add(msg)
+        }
+
+        companion object {
+            private val message = LinkedBlockingDeque<Message>()
+
+            fun startRecordingMessage() {
+                message.clear()
+            }
+
+            fun stopRecordingMessages(expectedElements: Int): List<Message?> {
```

Comment:
The naming and implementation are confusing. This functions doesn't just stop. It also tries to read the given number of messages, and will presumably throw due to timeouts on waiting for the right number to arrive. So what's really happening is it tries to return some recorded data, implicitly asserting that all requested messages exist.